internal/pg/postgres/password/scram.go

```diff
@@ -0,0 +1,259 @@
+	serverKey := s.hmac(scramDefaultHash, saltedPassword, scramServerKeyMessage)
+
+	// get the stored key, which is the hash of the client key
+	storedKey := s.hash(scramDefaultHash, clientKey)
+
+	// finally, we can build the scram verified!
+	verifier := fmt.Sprintf(scramVerifierFormat,
+		s.Iterations, s.encode(salt), s.encode(storedKey), s.encode(serverKey))
+
+	return verifier, nil
+}
+
+// encode creates a baseb64 encoding of a value that's returned as a string
+func (s *SCRAMPassword) encode(value []byte) string {
+	return base64.StdEncoding.EncodeToString(value)
+}
+
+// hash creates a SHA hash. Uses SHA-256, but can be swapped
+// in the future to use another hashing algorithm
+func (s *SCRAMPassword) hash(h func() hash.Hash, message []byte) []byte {
+	hf := h()
+	// hash.Hash.Write() never returns an error
+	_, _ = hf.Write(message)
+	return hf.Sum(nil)
+}
+
+// hmac performs a HMAC on a particular value. Uses SHA-256, but can be swapped
+// in the future to use another hashing algorithm
+func (s *SCRAMPassword) hmac(h func() hash.Hash, key, message []byte) []byte {
+	hm := hmac.New(h, key)
+	// hash.Hash.Write() never returns an error
+	_, _ = hm.Write(message)
+	return hm.Sum(nil)
+}
+
+// isASCII returns true if the string that is passed in is composed entirely of
+// ASCII characters
+func (s *SCRAMPassword) isASCII() bool {
+	// iterate through each character of the plaintext password and determine if
+	// it is ASCII. if it is not ASCII, exit early
+	// per research, this loop is optimized to be fast for searching
+	for i := 0; i < len(s.password); i++ {
+		if s.password[i] > unicode.MaxASCII {
+			return false
+		}
+	}
+
+	return true
+}
+
+// saslPrep returns the canonical form of a password for PostgreSQL when
+// using SCRAM. It differs from RFC 4013 in that it returns the original,
+// unmodified password when:
+//
+//  - the input is not valid UTF-8
+//  - the output would be empty
+//  - the output would contain prohibited characters
+//  - the output would contain ambiguous bidirectional characters
+//
+// See:
+//
+// https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/common/saslprep.c
+func (s *SCRAMPassword) saslPrep() string {
+	// if the password is only ASCII or it is not a valid UTF8 password, return
+	// the original password here
+	if s.isASCII() || !utf8.ValidString(s.password) {
+		return s.password
+	}
+
+	// perform SASLprep on the password. if the SASLprep fails or returns an
+	// empty string, return the original password
+	// Otherwise return the clean pasword
+	if cleanedPassword, err := stringprep.SASLprep.Prepare(s.password); cleanedPassword == "" || err != nil {
+		return s.password
+	} else {
+		return cleanedPassword
+	}
+}
+
+// setAttributes accepts just a plaintext password, and uses that to set the
+// password attribute. This also sets the default number of PBKDF2 iterations
+// and generated salt length
+func (s *SCRAMPassword) setAttributes(attr ...string) error {
```

Comment:
Is there a reason for this to be a variadic function?

Reply:
To match the interface. MD5 takes `username` + `password` whereas SCRAM only needs `password`.

I'm open to a better suggestion.

Reply:
Per @cbandy comment, I'm going to remove this and just use a constructor.